Add extra information around certificates (#915)
citusdata/citus_docs: admin_guide/cluster_management.rst

Connection Management
---------------------

.. note::

   Since Citus version 8.1.0 (released 2018-12-17) the traffic between the different nodes in the cluster is encrypted for NEW installations. This is done by using TLS with self-signed certificates. This means that this **does not protect against Man-In-The-Middle attacks.** This only protects against passive eavesdropping on the network.

>    For setting up self-signed TLS on installations that were originally created before Citus version 8.1.0 follow the steps in `official postgres documentation <https://www.postgresql.org/docs/current/ssl-tcp.html#SSL-CERTIFICATE-CREATION>`_ together with the citus specific settings described here. Setup should be done on coordinator and workers.

Reviewer: ```suggestion
   Clusters originally created with a Citus version before 8.1.0 do not have any network encryption enabled between nodes (even if upgraded later). To set up self-signed TLS on on this type of installation follow the steps in `official postgres documentation <https://www.postgresql.org/docs/current/ssl-tcp.html#SSL-CERTIFICATE-CREATION>`_ together with the citus specific settings described here, i.e. changing ``citus.node_conninfo`` to ``sslmode=require``. This setup should be done on coordinator and workers.
```

Lastly, to prevent any user from connecting via an unencrypted connection changes need to be made to ``pg_hba.conf``. Many Postgres installations will have entries allowing ``host`` connections which both allow SSL/TLS connections as well as plain TCP connections. By changing all ``host`` entries with ``hostssl`` entries only encrypted connections will be allowed to authenticate to Postgres. For full documentation on these settings take a look at `the pg_hba.conf file <https://www.postgresql.org/docs/current/auth-pg-hba-conf.html>`_ documentation on the official Postgres documentation.

.. note::

   When a trusted Certificate Authority is not available one can create their own via a self-signed root certificate. This is non-trivial and the developer or operator should seek guidance from their security team when doing so.

To verify the connections from the coordinator to the workers are encrypted you can run the following query. It will show the SSL/TLS version used to encrypt the connection that the coordinator uses to talk to the worker:

.. code-block:: postgresql

  SELECT run_command_on_workers($$
    SELECT version FROM pg_stat_ssl WHERE pid = pg_backend_pid()
  $$);
  

>   ┌────────────────────────────┐

Reviewer 2: Sphinx says: "Could not lex literal_block as "postgresql". Highlighting skipped."

I'd advise splitting these into two adjacent code blocks, like

```rst
.. code-block:: postgresql

   SELECT foo;

::

   output
```

Next to these machine specific files you need these cluster or CA wide files available:
 * ``/path/to/ca.crt``: Certificate of the Certificate Authority
 * ``/path/to/ca.crl``: Certificate Revocation List of the Certificate Authority

.. note::

   The Certificate Revocation List is likely to change over time. Work with your security department on a mechanism to get up to date versions of the revocation list on to all nodes in the cluster in a timely manner. A reload of every node in the cluster is required after the revocation list has been updated.

Once all files are in place on the nodes the following settings need to be configured in the Postgres configuration file:

.. code-block:: ini

   # the following settings enable the postgres server to enable ssl and configure the 
   # server to present the certificate to clients when connecting over tls/ssl
   ssl = on
   ssl_key_file = '/path/to/server.key'
   ssl_cert_file = '/path/to/server.crt'

   # this will tell citus to verify the certificate of the server it is connecting to 
   citus.node_conninfo = 'sslmode=verify-full sslrootcert=/path/to/ca.crt sslcrl=/path/to/ca.crl'

> After changing either restart the database or reload the configuration.

Reviewer: ```suggestion
After changing either restart the database or reload the configuration (a restart is required if a Citus version below 9.2.4 is used).
```

  ALTER DATABASE foo
  SET citus.node_conninfo =

>     'sslrootcert=/path/to/citus-ca.crt sslcrl=/path/to/citus-ca.crl sslmode=verify-full';

Author: added a section on how to setup with certificates from a trusted CA. Containing a note that one can setup a CA them self in the virtue of a self-signed root certificate.

>   ALTER DATABASE foo

Reviewer: What was the reason we have `ALTER DATABASE foo` here. During testing I always used `ALTER SYSTEM`. Does it make sense to change that here?

Author: This was already in here, @begriffs is best to answer the question on historical reasons I guess.

Reviewer 2: There's no deep history -- I introduced it in 626fab09. If alter system is preferred then it's fine to use that.

Reviewer: Okay, pushed commit that changes it to `ALTER SYSTEM`. That way we're sure it works.

  SET citus.node_conninfo =
    'sslrootcert=/path/to/citus-ca.crt sslcrl=/path/to/citus-ca.crl sslmode=verify-full';

> There is a whitelist of parameters that the GUC accepts, see the :ref:`node_conninfo <node_conninfo>` reference for details. As of Citus 8.1, the default value for node_conninfo is ``sslmode=require``, which prevents unencrypted communication between nodes.

Reviewer: ```suggestion
There is a whitelist of parameters that the GUC accepts, see the :ref:`node_conninfo <node_conninfo>` reference for details. As of Citus 8.1, the default value for node_conninfo is ``sslmode=require``, which prevents unencrypted communication between nodes. If your cluster was originally created before Citus 8.1 the value will be ``sslmode=prefer``. After setting up self-signed certificates on all nodes it's recommended to change this setting to ``sslmode=require``.
```

After changing this setting it is important to reload the postgres configuration. Even though the changed setting might be visible in all sessions, the setting is only consulted by Citus when new connections are established. When a reload signal is received citus marks all existing connections to be closed which causes a reconnect after running transactions have been completed.

.. code-block:: postgresql

>   SELECT pg_reload_conf();

Reviewer 2: Is this something that needs to be propagated through `run_command_on_workers()`?

Reviewer: Good question. That depends a bit what you mean. If you only change `citus.node_conninfo` on one node, there's no reason to reload the configuration on the other nodes. Often you probably want to change the configuration on all nodes. So then it would make sense to do something like:
```sql
ALTER SYSTEM SET citus.node_conninfo =
    'sslrootcert=/path/to/root.crt sslcrl=/path/to/root.crl sslmode=verify-full';
select pg_reload_conf();
select run_commands_on_workers($$
ALTER SYSTEM SET citus.node_conninfo =
    'sslrootcert=/path/to/root.crt sslcrl=/path/to/root.crl sslmode=verify-full';
$$)
select run_commands_on_workers($$
select pg_reload_conf();
$$)
```

This assumes these files are in the same place on all nodes though.

I think it's fine to leave this like this for the initial version of this change.

.. warning:: 

>    Versions before 9.2.4 require a restart for existing connections to be closed.

Reviewer: ```suggestion
   Citus versions before 9.2.4 require a restart for existing connections to be closed.
```

For all nodes in the cluster you need to get a valid certificate signed by the *same Certificate Authority*. The following **machine specific** files are assumed to be available on every machine:
 * ``/path/to/server.key``: Server Private Key
 * ``/path/to/server.crt``: Server Certifiacte or Certificate Chain for Server Key, signed by trusted Certifiacte Authority.

Next to these machine specific files you need these cluster or CA wide files available:
 * ``/path/to/ca.crt``: Certificate of the Certificate Authority
 * ``/path/to/ca.crl``: Certificate Revocation List of the Certificate Authority

.. note::

   The Certificate Revocation List is likely to change over time. Work with your security department on a mechanism to get up to date versions of the revocation list on to all nodes in the cluster in a timely manner. A reload of every node in the cluster is required after the revocation list has been updated.

Once all files are in place on the nodes the following settings need to be configured in the Postgres configuration file:

.. code-block:: ini

   # the following settings enable the postgres server to enable ssl and configure the server to present the certifacte

>    # to clients when connecting over tls/ssl

Reviewer: This comment was not fully readable in the default layout without scrolling right
```suggestion
   # the following settings enable the postgres server to enable ssl and configure the 
   # server to present the certificate to clients when connecting over tls/ssl
```

> Once all files are in place on the nodes the following settings need to be configured in the Postgres configuration file:

Reviewer: ```suggestion
Once all files are in place on the nodes the following settings need to be configured in the Postgres configuration file (PostgreSQL needs to be restarted for these settings to be applied):
```

Author: I don't think postgres requires a restart. I believe ssl changes are reloaded in postgres since pg10. I think adding a reload is required makes sense.

Reviewer: Makes sense, I updated my commit suggestion accordingly.

Author: I have more generally phrased it and put it below the config block instead. Do you think it is ok like that now?

Reviewer: I added a small comment updating that wording. I think with that it's okay.

   ssl = on
   ssl_key_file = '/path/to/server.key'
   ssl_cert_file = '/path/to/server.crt'

   # this will tell citus to verify the certificate of the server it is connecting to 

>    citus.node_conninfo = 'sslmode=verify-ca sslrootcert=/path/to/ca.crt sslcrl=/path/to/ca.crl'

Reviewer: When doing this on a running cluster you will want to split up the `ssl=on` and the `citus.node_conninfo` changes. Otherwise you will get downtime, because not all servers are using the new keys/certificates when accepting connections.

Author: I don't think the documentation becomes more clear by adding this layer on top in the Security section.

@begriffs do we have documentation for zero downtime upgrades? Maybe we can fold this information in there.

Reviewer: Makes sense, this is not required for this for this first version anyway. I did leave one more comment, which notes that you need to restart PostgreSQL for these settings to apply.

Reviewer 2: We have a section about [upgrading citus](https://docs.citusdata.com/en/v9.2/admin_guide/upgrading_citus.html) itself, but it doesn't talk about downtime. We could add a section in there.

Author: I think that makes more sense to have a sprint at some point where we pay more attention to the zero downtime operations part of citus.

zero downtime is frankly not a security concern :)

> Depending on the policy of the Certificate Authority used you might need or want to change ``sslmode=verify-ca`` in ``citus.node_conninfo`` to ``sslmode=verify-full``. For the difference between the two settings please consult `the official postgres documentation <https://www.postgresql.org/docs/current/libpq-ssl.html#LIBPQ-SSL-SSLMODE-STATEMENTS>`_.

Reviewer: Let's recommend the most secure option.
```suggestion
   citus.node_conninfo = 'sslmode=verify-full sslrootcert=/path/to/ca.crt sslcrl=/path/to/ca.crl'

Depending on the policy of the Certificate Authority used you might need or want to change ``sslmode=verify-full`` in ``citus.node_conninfo`` to ``sslmode=verify-ca``. For the difference between the two settings please consult `the official postgres documentation <https://www.postgresql.org/docs/current/libpq-ssl.html#LIBPQ-SSL-SSLMODE-STATEMENTS>`_.
```
